src/main/java/org/reaktivity/nukleus/oauth/internal/stream/OAuthProxyFactory.java

```diff
@@ -245,6 +261,64 @@ private RouteFW wrapRoute(
         return routeRO.wrap(buffer, index, index + length);
     }
 
+    private OAuthAccessGrant modifyGrant(
+        JsonWebSignature verified,
+        long affinity,
+        long connectAuthorization,
+        long expiresAtMillis)
+    {
+        OAuthAccessGrant grant = null;
+        try
+        {
+            if(verified != null)
+            {
+                final JwtClaims claims = JwtClaims.parse(verified.getPayload());
+                final String subject = claims.getSubject();
+                if(subject != null)
+                {
+                    final int realmId = (int) ((connectAuthorization & REALM_MASK) >> SCOPE_BITS);
+
+                    grant = supplyGrant(realmId, affinity, subject);
+
+                    if(grant.referenceCount > 0)
+                    {
+                        final long grantAuthorization = grant.authorization;
+
+                        if((grantAuthorization & connectAuthorization) == grantAuthorization && expiresAtMillis > grant.expiresAt)
+                        {
+                            grant.expiresAt = expiresAtMillis;
+                        }
+                    }
+                    else if(grant.referenceCount == 0)
+                    {
+                        grant.subject = subject;
+                        grant.authorization = connectAuthorization;
+                        grant.expiresAt = expiresAtMillis;
+                    }
```

Comment:
Move this `if-else` block to a method on `OAuthAccessGrant`, perhaps called `reauthorize` returning `boolean` if successfully reauthorized according to enforced compatibility conditions.